src/main/java/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep.java

```diff
@@ -56,16 +91,29 @@
          */
         private final String id = UUID.randomUUID().toString();
         private static final float RECURRENCE_PERIOD_BACKOFF = 1.2f;
-        static final long MIN_RECURRENCE_PERIOD = 250; // ¼s
-        static final long MAX_RECURRENCE_PERIOD = 15000; // ¼min
-        long recurrencePeriod = MIN_RECURRENCE_PERIOD;
+        long recurrencePeriod;
+        long maxRecurrencePeriod;
+        long minRecurrencePeriod;
+        private transient final WaitForConditionStep step;
 
-        Execution(StepContext context) {
+        Execution(WaitForConditionStep step, StepContext context) {
             super(context);
+            this.step = step;
+        }
+        
+        private Object readResolve() {
+            minRecurrencePeriod = DescriptorImpl.DEFAULT_MIN_RECURRENCE_PERIOD;
+            maxRecurrencePeriod = DescriptorImpl.DEFAULT_MAX_RECURRENCE_PERIOD;
+            return this;
         }
 
         @Override public boolean start() throws Exception {
             body = getContext().newBodyInvoker().withCallback(new Callback(id)).start();
+            this.recurrencePeriod = step.unit.toMillis(step.getMinRecurrencePeriod());
```

Comment:
Would `step.unit` be null here for running builds if we upgraded the plugin and then restarted  Jenkins?